routes/userSubmissionRoutes.js

```diff
@@ -0,0 +1,72 @@
+const express = require('express');
+const config = require('../config');
+const multer = require('multer');
+const parts = multer();
+const wrap = require('co-express');
+const co = require('co');
+const QuestionController = require('../controllers/questionController');
+const TestQuestions = require('../testObjects/TestQuestions');
+
+const RunnerController = require('../controllers/runnerController');
+
+module.exports = function (app) {
+
+    app.post('/user_submission', parts.array(), wrap(function *(req, res) {
+        const questionId = req.body.questionId;
+        const userName = req.body.userId;
+        const userCode = req.body.userCode;
+
+        co(function *() {
+
+            // get question from the database
+            var question = yield QuestionController.findById(questionId);
+
+            // If the outputs of the complete solution are not stored
+            // run them now and store in the database
+            if (!question.testCases[0].output) {
+                yield RunnerController.compileAndRunTestCases(
+                    question, "dummyUser", "", true);
+                yield QuestionController.save(question);
+            }
+
+            // Compile and run test cases for question
+            var runStatus = yield RunnerController.compileAndRunTestCases(
+                question, userName, userCode, false);
+
+            // Determine which test cases are correct
+            var testCases = question.testCases;
+            var results = [];
+            var numCorrect = 0;
+            for (var i=0; i<runStatus.length; i++) {
+                if (testCases[i].output == runStatus[i]) {
+                    results.push(true);
+                    numCorrect++;
+                }
+                else {
+                    results.push(false);
+                }
+            }
+            var score = question.points * (numCorrect / question.testCases.length);
+
+            // Create User Submission object
+            var userSubmission = {
+                username: userName,
+                question: questionId,  // ?????????? Can this be a String
```

Comment:
It depends. With the current schema it can not. In the current schema we have it setup as reference then this needs to be a Question itself rather than just the ID